java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java

```diff
@@ -19,223 +19,69 @@
-      case RETURN_STATEMENT:
-        Tree method = parent;
-        while (!method.is(Tree.Kind.METHOD, Tree.Kind.LAMBDA_EXPRESSION)) {
-          method = method.parent();
-        }
-        if (method.is(Tree.Kind.METHOD)) {
-          return ((MethodTree) method).symbol().returnType().type();
-        }
-        return null;
-      case VARIABLE:
-        return ((VariableTree) parent).symbol().type();
-      case ARGUMENTS:
-        Arguments arguments = (Arguments) parent;
-        Tree invocation = arguments.parent();
-        if (invocation.is(Tree.Kind.METHOD_INVOCATION)) {
-          return targetTypeFromMethodSymbol(((MethodInvocationTree) invocation).symbol(), arguments, typeCastTree);
-        }
-        if (invocation.is(Tree.Kind.NEW_CLASS)) {
-          return targetTypeFromMethodSymbol(((NewClassTree) invocation).constructorSymbol(), arguments, typeCastTree);
-        }
-        // Last possible case is AnnotationTree which we do not consider
-        return null;
-      case MEMBER_SELECT:
-      case CONDITIONAL_EXPRESSION:
-        return typeCastTree.type().symbolType();
-      case ARRAY_ACCESS_EXPRESSION:
-        return ((ArrayAccessExpressionTree) parent).expression().symbolType();
-      default:
-        if (parent instanceof ExpressionTree) {
-          return ((ExpressionTree) parent).symbolType();
-        }
-        return null;
+  private boolean isUnnecessaryCast(TypeCastTree typeCastTree) {
+    if (skipParentheses(typeCastTree.expression()).is(Tree.Kind.NULL_LITERAL)) {
+      Tree parentTree = skipParentheses(typeCastTree.parent());
+      return !parentTree.is(Tree.Kind.ARGUMENTS);
     }
+    return warnings.stream().anyMatch(warning -> matchesWarning(warning, typeCastTree));
   }
 
-  @CheckForNull
-  private static Type targetTypeFromMethodSymbol(Symbol symbol, Arguments arguments, TypeCastTree typeCastTree) {
-    if (symbol.isMethodSymbol()) {
-      Symbol.MethodSymbol sym = (Symbol.MethodSymbol) symbol;
-      int castArgIndex = indexOfTypeCast(arguments, typeCastTree);
-      List<Type> parameterTypes = sym.parameterTypes();
-      int nbParameters = parameterTypes.size();
-      // Handle varargs
-      return nbParameters > castArgIndex ? parameterTypes.get(castArgIndex) : parameterTypes.get(nbParameters - 1);
-    }
-    return null;
-  }
-
-  private static int indexOfTypeCast(Arguments arguments, TypeCastTree typeCastTree) {
-    int i = 0;
-    do {
-      // typeCastTree is necessarily within the arguments
-      Tree arg = ExpressionUtils.skipParentheses(arguments.get(i));
-      if (typeCastTree.equals(arg)) {
-        return i;
+  private static boolean matchesWarning(JWarning warning, TypeCastTree tree) {
+    SyntaxToken startToken = tree.openParenToken();
+    // When a cast expression is nested inside one or more parenthesized expression, Eclipse raises the warning on
+    // the outermost parenthesized expression rather than the cast expression, so we need to take that into account
```

Comment:
I assume this also applies to other Eclipse warnings that are raised on expressions and should be taken into account when we implement the logic of mapping warnings to trees.